File: app/src/main/java/top/maweihao/weather/view/BaseActivity.java
package top.maweihao.weather.view;

import android.os.Bundle;
import android.preference.PreferenceManager;

import com.trello.rxlifecycle2.components.support.RxAppCompatActivity;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import top.maweihao.weather.R;

public abstract class BaseActivity extends RxAppCompatActivity {

    protected final String TAG = this.getClass().getSimpleName();

    private Unbinder bind;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(setContentView());
        bind = ButterKnife.bind(this);
        PreferenceManager.setDefaultValues(this, R.xml.settingpreference, false);

        initView(savedInstanceState);
    }

    /**
     * 子类设置界面
     */
    protected abstract int setContentView();

    protected abstract void initView(Bundle savedInstanceState);

    @Override
    protected void onDestroy() {
        super.onDestroy();
        bind.unbind();
    }
}

File: app/src/main/java/top/maweihao/weather/view/BaseActivity.kt
package top.maweihao.weather.view

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

abstract class BaseActivity : AppCompatActivity() {

    protected val TAG = this.javaClass.simpleName

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(setContentView())

        initView(savedInstanceState)
        initData()
    }

    /**
     * 子类设置界面
     */
    protected abstract fun setContentView(): Int

    protected abstract fun initView(savedInstanceState: Bundle?)

    protected abstract fun initData()

}
